src/device_trezor/device_trezor_base.cpp

```diff
@@ -269,32 +277,65 @@ namespace trezor {
       return false;
     }
 
-    void device_trezor_base::on_button_request()
+    void device_trezor_base::on_button_request(GenericMessage & resp, const messages::common::ButtonRequest * msg)
     {
+      CHECK_AND_ASSERT_THROW_MES(msg, "Empty message");
+      MDEBUG("on_button_request, code: " << msg->code());
+
+      messages::common::ButtonAck ack;
+      write_raw(&ack);
+
       if (m_callback){
         m_callback->on_button_request();
       }
+
+      resp = read_raw();
     }
 
-    void device_trezor_base::on_pin_request(epee::wipeable_string & pin)
+    void device_trezor_base::on_pin_request(GenericMessage & resp, const messages::common::PinMatrixRequest * msg)
     {
+      MDEBUG("on_pin_request");
+      CHECK_AND_ASSERT_THROW_MES(msg, "Empty message");
+
+      epee::wipeable_string pin;
+
       if (m_callback){
         m_callback->on_pin_request(pin);
       }
+
+      messages::common::PinMatrixAck m;
+      m.set_pin(pin.data(), pin.size());
```

Comment:
This will presumably leave that data in memory when done. I'm guessing fixing that isn't trivial so OK to go in, but maybe make a note of it for later.

Reply:
This is thought one indeed. Protobuf generated code uses `std::string` to represent these fields. I will add comments here. It may be needed to change the Trezor protocol to use byte arrays for secrets in the future.